src/engine/IdTable.h

```diff
@@ -0,0 +1,933 @@
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+  const_iterator end() const {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+  const_iterator cend() const {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+
+  Row back() {
+    return Row(IdTableImpl<COLS>::_data +
+                   (IdTableImpl<COLS>::_size - 1) * IdTableImpl<COLS>::_cols,
+               IdTableImpl<COLS>::_cols);
+  }
+
+  ConstRow back() const {
+    return ConstRow(IdTableImpl<COLS>::_data + (IdTableImpl<COLS>::_size - 1) *
+                                                   IdTableImpl<COLS>::_cols,
+                    IdTableImpl<COLS>::_cols);
+  }
+
+  Id* data() { return IdTableImpl<COLS>::_data; }
+  const Id* data() const { return IdTableImpl<COLS>::_data; }
+
+  // Size access
+  size_t rows() const { return IdTableImpl<COLS>::_size; }
+
+  /**
+   * The template parameter here is used to allow for creating
+   * a version of the function that is constexpr for COLS != 0
+   * and one that is not constexpt for COLS == 0
+   */
+  using IdTableImpl<COLS>::cols;
+
+  size_t size() const { return IdTableImpl<COLS>::_size; }
+
+  /**
+   * @brief Sets the number of columns. Should only be called while data is
+   *        still nullptr.
+   **/
+  void setCols(size_t cols) {
+    assert(IdTableImpl<COLS>::_data == nullptr);
+    IdTableImpl<COLS>::setCols(cols);
+  }
+
+  void emplace_back() { push_back(); }
+
+  void push_back() {
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    IdTableImpl<COLS>::_size++;
+  }
+
+  void push_back(const std::initializer_list<Id>& init) {
+    assert(init.size() == IdTableImpl<COLS>::_cols);
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
+                    IdTableImpl<COLS>::_size * IdTableImpl<COLS>::_cols,
+                init.begin(), sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  /**
+   * @brief Read cols() elements from init and stores them in a new row
+   **/
+  void push_back(const Id* init) {
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
+                    IdTableImpl<COLS>::_size * IdTableImpl<COLS>::_cols,
+                init, sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  /**
+   * @brief Read cols() elements from init and stores them in a new row
+   **/
+  void push_back(const Row& init) {
+    assert(init.cols() == cols());
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
```

Comment:
I think a normal assignment might be preferable to forcing a `memcpy` here?

Reply:
Using `memcpy` avoids the less efficient row based access here (which would construct and object). As this is part of the internals of the `IdTable` I went for efficiency over simplicity here.